In [199]:
import pandas as pd

DATA_PATH_RUI = './data/bq-results-20190710-040500-jn11labq5atn.json'
DATA_PATH_SIM = './data/bq-results-20190710-041233-sg10wldt6l6m.json'

def read_jsonl(path):
    with open(path, 'r') as f:
        data = f.readlines()
        
    data = map(lambda x: x.rstrip(), data)
    data_json_str = "[" + ','.join(data) + "]"
    return pd.read_json(data_json_str)

df_rui = read_jsonl(DATA_PATH_RUI).query('tournament=="NCAA"')
df_sim = read_jsonl(DATA_PATH_SIM)

print(df_rui.columns)


Index(['abbr_name', 'active', 'assists', 'assists_turnover_ratio',
       'birth_place', 'birthplace_city', 'birthplace_country',
       'birthplace_state', 'blocked_att', 'blocks', 'class', 'conf_alias',
       'conf_name', 'defensive_rebounds', 'division_alias', 'division_name',
       'field_goals_att', 'field_goals_made', 'field_goals_pct', 'first_name',
       'flagrant_fouls', 'free_throws_att', 'free_throws_made',
       'free_throws_pct', 'full_name', 'game_id', 'gametime', 'height',
       'home_team', 'jersey_number', 'last_name', 'league_name', 'minutes',
       'minutes_int64', 'neutral_site', 'offensive_rebounds', 'personal_fouls',
       'played', 'player_id', 'points', 'position', 'primary_position',
       'rebounds', 'scheduled_date', 'season', 'sp_created', 'starter',
       'status', 'steals', 'team_alias', 'team_id', 'team_market', 'team_name',
       'tech_fouls', 'three_points_att', 'three_points_made',
       'three_points_pct', 'tournament', 'tournament_game_no'

In [200]:
names = df_sim.abbr_name.unique()
print(f'The number of similar players to Rui Hachimura = {len(names)}')

The number of similar players to Rui Hachimura = 1300


In [201]:
str_labels = [
    # プレイヤーID
    'player_id',
    
    # プレイヤー名
    'full_name',
    
    # シーズン
    'season',
]

num_labels = [
    # アシスト数
    'assists',

    # ブロック数
    'blocks',
    
    # ディフェンスリバウンド数
    'defensive_rebounds',

    # フィールドゴール試投数
    'field_goals_att',
    
    # フィールドゴール成功数
    'field_goals_made',
    
    # フリースロー試投数
    'free_throws_att',
    
    # フリースロー成功数
    'free_throws_made',
    
    # 身長
    'height',
    
    # 出場時間
    'minutes',
    
     # オフェンスリバウンド数
    'offensive_rebounds',
    
    # 得点数
    'points',
    
    # リバウンド数
    'rebounds',
    
    # スティール数
    'steals',
    
    # 3P試投数
    'three_points_att',
    
    # 3P成功数
    'three_points_made',
    
    # ターンオーバー数
    'turnovers',
    
    # 2P試投数
    'two_points_att',
    
    # 2P成功数
    'two_points_made',
    
    # 体重
    'weight',
]

# 2018年のデータがないので手動で追加
# https://www.espn.com/mens-college-basketball/team/stats/_/id/2250
rui_2018 = {
    'abbr_name': ' R.Hachimura',
    'active': 1.0,
    'assists': 1.5,
    'blocked_att': None,
    'blocks': 0.7,
    'defensive_rebounds': 188/37,
    'field_goals_att': 465/37,
    'field_goals_made': 275/37,
    'field_goals_pct': 59.1,
    'first_name': "Rui",
    'flagrant_fouls': None,
    'free_throws_att': 222/37,
    'free_throws_made': 164/37,
    'free_throws_pct': 73.9,
    'full_name': "Rui Hachimura",
    'height': 80,
    'last_name': "Hachimura",
    'minutes': 30.2,
    'offensive_rebounds': 52/37,
    'personal_fouls': None,
    'player_id': '66a86e68-5bfc-4e7a-9610-1a7a7ae406b4',
    'points': 19.7,
    'position': "F",
    'primary_position': "NA",
    'rebounds': 6.5,
    'season': 2018,
    'starter': True,
    'status': "ACT",
    'steals': 0.9,
    'team_id': '2f4d21f8-6d5f-48a5-abca-52a30583871a',
    'team_market': 'Gonzaga',
    'team_name': 'Bulldogs',
    'tech_fouls': None,
    'three_points_att': 36/37,
    'three_points_made': 15/37,
    'three_points_pct': 41.7,
    'tournament': "NCAA",
    'turnovers': 1.8,
    'two_points_att': (465-36)/37,
    'two_points_made': (275-15)/37,
    'two_points_pct': 59.1,
    'weight': 225,
}
df_rui = pd.DataFrame([rui_2018])

In [202]:
def convert_minutes(val):
    if type(val) is int:
        return float(val)
    elif type(val) is float:
        return val
    elif type(val) is str:
        time = val.split(':')
        return float(time[0]) + float(time[1]) / 60.0
    else:
        return 0.0

df_rui['minutes'] = df_rui.minutes.apply(convert_minutes)
df_sim['minutes'] = df_sim.minutes.apply(convert_minutes)

In [203]:
import numpy as np
from sklearn.neighbors import NearestNeighbors

df_dummy = df_sim.loc[:, num_labels].dropna()
knn = NearestNeighbors(n_neighbors=3, algorithm='ball_tree')
knn.fit(df_dummy)

NearestNeighbors(algorithm='ball_tree', leaf_size=30, metric='minkowski',
                 metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                 radius=1.0)

In [204]:
def is_number(x):
    return (type(x) is int or type(x) is float) and not np.isnan(x)


def neighbor_or_mean(df, label, neighbors):
    mean = df[label].dropna().mean()

    def f(row):        
        v = None
        if is_number(row[label]):
             v = row[label]

        else:
            for neigh in neighbors:
                x = df.loc[neigh, [label]]
                if is_number(x):
                    v = x
                    break

            # 最近傍のデータが値を持たなければ平均値を使用する
            if v is None:
                v = mean
        return v    
    return f


def fillna(df):
    _df = df.copy()
    
#     _, neighbors = knn.kneighbors(_df.loc[:, num_labels].fillna(0.0))
    neighbors = []

    for label in num_labels:
        _df[label] = df.apply(neighbor_or_mean(df, label, neighbors), axis=1)
    return _df


_df_sim = fillna(df_sim)
df_sim = _df_sim

In [208]:
df_rui_stats = df_rui.loc[:, labels].groupby(['player_id', 'season', 'full_name']).mean()
df_sim_stats = df_sim.loc[:, labels].groupby(['player_id', 'season', 'full_name']).mean()

In [209]:
df_rui_stats.head(1)

assists  blocks  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura      1.5     0.7   

                                                           defensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            5.081081   

                                                           field_goals_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura        12.567568   

                                                           field_goals_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          7.432432   

                                                           free_throws_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura              6.0   

                                                           free_throws_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          4.432432   

                                                           height  minutes  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura      80     30.2   

                                                           offensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            1.405405   

                                                           points  rebounds  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura    19.7       6.5   

                                                           starter  steals  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura     True     0.9   

                                                           three_points_att  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          0.972973   

                                                           three_points_made  \
player_id                            season full_name                          
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura           0.405405   

                                                           turnovers  \
player_id                            season full_name                  
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura        1.8   

                                                           two_points_att  \
player_id                            season full_name                       
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura       11.594595   

                                                           two_points_made  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         7.027027   

                                                           weight  
player_id                            season full_name              
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura     225

In [210]:
df_sim_stats.head(3)

assists  \
player_id                            season full_name                   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith       1.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald  0.185062   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov  0.000000   

                                                              blocks  \
player_id                            season full_name                  
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith       2.00000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald  0.12826   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov  0.00000   

                                                             defensive_rebounds  \
player_id                            season full_name                             
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith                 4.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald            0.637154   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov            0.000000   

                                                             field_goals_att  \
player_id                            season full_name                          
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith              6.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald         1.227746   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov         0.000000   

                                                             field_goals_made  \
player_id                            season full_name                           
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith               4.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald          0.588134   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov          0.000000   

                                                             free_throws_att  \
player_id                            season full_name                          
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith              2.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald         0.444815   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov         0.000000   

                                                             free_throws_made  \
player_id                            season full_name                           
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith               0.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald          0.302608   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov          0.000000   

                                                             height  \
player_id                            season full_name                 
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith           80   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald      78   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov      80   

                                                               minutes  \
player_id                            season full_name                    
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith       34.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald  17.486401   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov  17.486401   

                                                             offensive_rebounds  \
player_id                            season full_name                             
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith                 0.000000   
00637b09-c75d-438a-82c7-02467546bbfb 2015   Jaylin McDonald            0.311306   
00ae065b-9c78-49b8-b9cf-5d8872ef2d1d 2014   Georgi Funtarov            0.000000   

                                                               points  \
player_id                            season full_name                   
0060a1bd-1c7b-43

In [211]:
# dataframe を結合
df = pd.concat([df_rui_stats.loc[:, num_labels], df_sim_stats.loc[:, num_labels]])

In [212]:
# 各値を標準化（N(0, 1) に従う分布に変換）する
def standardize(df, labels):
    _df = df.copy()
    for label in labels:
        mean = df[label].mean()
        std = df[label].std()
        values = []
        for val in df[label].values:
            if std > 0:
                values.append((val - mean) / std)
            else:
                values.append(0.0)
        _df[label] = values
    return _df

df = standardize(df, num_labels)

In [213]:
# 各値を [0, 1] に正規化
def normalize(df, labels):
    _df = df.copy()
    for label in labels:
        max_val = df[label].max()
        min_val = df[label].min()
        values = []
        for val in df[label].values:
            if max_val != min_val:
                values.append((val - min_val) / (max_val - min_val))
            else:
                values.append(0.0)
        _df[label] = values
    return _df

# df = normalize(df, num_labels)

In [214]:
df.head(2)

assists    blocks  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  1.141035  0.525911   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     0.545136  2.558968   

                                                           defensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            1.523418   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith               1.009754   

                                                           field_goals_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         2.175980   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            0.560716   

                                                           field_goals_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          2.792317   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith             1.109352   

                                                           free_throws_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         2.506178   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            0.368754   

                                                           free_throws_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          2.588536   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            -0.642686   

                                                             height   minutes  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  0.170032  1.341872   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     0.170032  1.744487   

                                                           offensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            0.399980   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith              -0.789117   

                                                             points  rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  2.828208  1.245405   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     0.632685  0.402863   

                                                             steals  \
player_id                            season full_name                 
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  1.021130   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     1.201299   

                                                           three_points_att  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          0.106052   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            -0.535500   

                                                           three_points_made  \
player_id                            season full_name                          
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura           0.211940   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith             -0.438531   

                                                           turnovers  \
player_id                            season full_name                  
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura   1.179414   
0060a1bd-1c7b-43da-8690-63962bd28aa6

In [215]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=16)
kmeans.fit(df)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=16, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [216]:
df['cluster'] = kmeans.labels_

In [217]:
from sklearn.metrics.pairwise import cosine_similarity

def calc_cos(row, df):
    cosines = []
    for _, x in df.iterrows():
        X = row.values.reshape(1, -1)
        Y = x.values.reshape(1, -1)
        cosines.append(cosine_similarity(X, Y)[0][0])
    return cosines

cos = calc_cos(df.iloc[0], df)
df['cosine'] = cos

In [218]:
df.head(2)

assists    blocks  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  1.141035  0.525911   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     0.545136  2.558968   

                                                           defensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            1.523418   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith               1.009754   

                                                           field_goals_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         2.175980   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            0.560716   

                                                           field_goals_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          2.792317   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith             1.109352   

                                                           free_throws_att  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         2.506178   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            0.368754   

                                                           free_throws_made  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          2.588536   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            -0.642686   

                                                             height   minutes  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  0.170032  1.341872   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     0.170032  1.744487   

                                                           offensive_rebounds  \
player_id                            season full_name                           
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura            0.399980   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith              -0.789117   

                                                           ...  rebounds  \
player_id                            season full_name      ...             
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  ...  1.245405   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     ...  0.402863   

                                                             steals  \
player_id                            season full_name                 
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura  1.021130   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith     1.201299   

                                                           three_points_att  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          0.106052   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith            -0.535500   

                                                           three_points_made  \
player_id                            season full_name                          
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura           0.211940   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith             -0.438531   

                                                           turnovers  \
player_id                            season full_name                  
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura   1.179414   
0060a1bd-1c7b-43da-8690-63962bd28aa6 2015   Zach Smith  

In [219]:
no_cluster_hachimura = df.cluster.values[0]
print(f'cluster no. of Hachimura = {no_cluster_hachimura}')

df.query(f'cluster=={no_cluster_hachimura}')

cluster no. of Hachimura = 0


assists  \
player_id                            season full_name                         
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura          1.141035   
0bd137c0-85f7-4395-a51b-88f707a4f8d1 2015   Laimonas Chatkevicius  0.545136   
11702dc5-fd09-4b71-8c54-a08af0c91dd2 2015   Jaylen Bond           -0.646661   
146aed9c-0443-49e0-b537-9fa2669af16c 2014   Jarell Martin          4.120527   
1df2ebf8-f260-44e1-b41f-10739d1f02e5 2015   Ethan Happ             1.339667   
1f94bc17-7ea1-4fa2-b1a1-949f9ca1902e 2014   Savon Goodman          1.141035   
200b9d70-e2a7-4a43-84ad-80782480645d 2014   Chris Hooper           0.545136   
236e598c-831d-4b17-b121-5b9fe5881305 2015   Yante Maten            1.141035   
272b0157-c3cf-409e-80a6-9a7f7500a5b7 2014   Javonte Green          1.141035   
2af2ce51-5235-4403-9841-4c9430cb8f1b 2015   Matt Costello          1.736933   
2c84fa41-3835-41cc-baf6-438ebd2bf50f 2014   Dexter Werner         -0.646661   
30c49b9a-d7f7-4c35-8cd1-7f38babaa64a 2015   Twymond Howard         1.736933   
32cfb9a3-f83a-4a8b-b2bf-3a9b3c66c01a 2014   RaShid Gaston          4.120527   
33ed07ae-50b7-435e-b3a8-4ed53b86d780 2013   Dustin Hogue           1.339667   
343d8b00-c5c9-4e3d-9a17-6f2ad78a279a 2014   Emmy Andujar           0.545136   
388afda2-a885-460f-bc7a-8881626775b0 2013   Ronald Roberts, Jr.   -0.646661   
3d64c487-51d4-4d54-9d1a-e6e0a9e3d13b 2014   Josiah Heath           0.942402   
3e60a457-0e1d-4655-810a-3e0d8d9badb6 2014   Dan Trist              0.545136   
480f8b24-cc49-4b47-8d21-cfc4937e0d6f 2015   Martin Breunig         1.736933   
48a44e9f-d17d-46af-a730-b7563e63101e 2013   Perry Ellis           -0.646661   
4a1eec63-6427-4913-8981-ce4bf403bb10 2015   Ivan Rabb              1.736933   
557b959b-7883-4284-b56a-766d47464cd3 2015   Dion Wright            0.545136   
64f64d17-264d-4f13-b6fe-4b769ad0aa19 2015   Grandy Glaze           0.147870   
7259bc8b-f7a0-49e4-b4a6-86f52d88e903 2015   Brandon Sherrod        0.545136   
737a1703-71aa-445c-957e-64a77d1ebefe 2014   Jarrod Uthoff          0.545136   
7fd76e7e-c999-49ad-97f0-f85b8d569a0e 2015   Tanner Leissner        1.736933   
89e24b50-50e5-4434-a616-a4c1153ae8a4 2014   JJ O'Brien             1.141035   
8a8c45be-bce1-4adc-9b9e-f8a83c079d8b 2015   Pascal Siakam          0.545136   
92ce1de7-34a3-4a36-96e3-ba89fc2d79c7 2014   Earl Brown             2.928730   
991414c2-6bdc-4987-86ae-48080cdc947c 2013   Markus Kennedy         1.021855   
                                     2014   Markus Kennedy        -0.646661   
9e4b6d59-dfb3-4ebf-abb8-4d42485b86b6 2014   Venky Jois             0.545136   
9e708a5c-b597-44af-b883-79c4dfe83f63 2014   Jimmy Hall             0.942402   
a691c25f-d858-49dc-a744-afbd5cfe7912 2014   Will Davis II         -0.646661   
b40cd2a0-78bf-497a-bb4c-b4bc35c715a5 2013   James Michael McAdoo   2.332832   
b50bf001-245d-4f61-b42e-020c05ec953a 2014   Zach Auguste          -0.050762   
b7e39c95-4504-426b-ba65-847d1ff61a27 2014   Scott Eatherton       -0.646661   
c1539a69-c48d-4c51-8f88-22d035bf8071 2015   Lyonell Gaines         1.538300   
c382acd4-ad9b-47c9-b786-79eb0b6806b1 2014   Stanton Kidd           0.545136   
c5b8338d-bd73-4a5a-9be3-011b748dabfd 2013   Dwayne Evans          -0.646661   
c60137bd-db7b-4715-9c07-9d239eb5a1e2 2015   Marc-Eddy Norelia      2.928730   
c6f5ca6a-9003-4d48-882d-8d2bd2f255f3 2014   Garrett Jackson        0.545136   
d3c0d366-a44d-470f-ab0e-22ec3097a261 2014   Roosevelt Jones        3.524629   
e00cfdff-6eef-4d2c-886c-b1f3efaff0ff 2015   Michael Rowley         4.120527   
e4fa7ee2-b4e3-4b6f-9374-b01f2c6eb406 2013   C.J. Fair             -0.050762   
e8642127-2203-453c-be7d-69ea89f22cc1 2013   Isaiah Armwood         0.545136   
f4b819cd-9d99-4a88-867e-78142bb69198 2015   Justin Sears           2.332832   
f63de964-6a62-4c62-a317-9e74129bcc9f 2014   Jordyn Martin          0.783495   
fd02a07f-422f-44b3-8177-3b793cb40d28 2013   Doug McDermott         1.141035   
fd8a7b75-2f5f-49fa-8d85-9b93a926f821 20

In [220]:
similar_players = df.query(f'cluster=={no_cluster_hachimura}')

In [198]:
df_similar = df.sort_values('cosine', ascending=False)
df_similar

assists  \
player_id                            season full_name                        
66a86e68-5bfc-4e7a-9610-1a7a7ae406b4 2018   Rui Hachimura         1.083755   
5ddccd3b-97ed-4a08-92ac-77c6c1e28b8c 2015   Anthony Gill          1.369218   
48a44e9f-d17d-46af-a730-b7563e63101e 2015   Perry Ellis           0.227367   
2c4481c6-6c41-47a7-b9c6-f8ae3a54a185 2016   Chima Moneke         -0.058096   
33ed07ae-50b7-435e-b3a8-4ed53b86d780 2013   Dustin Hogue          1.274064   
b40cd2a0-78bf-497a-bb4c-b4bc35c715a5 2013   James Michael McAdoo  2.225606   
480f8b24-cc49-4b47-8d21-cfc4937e0d6f 2015   Martin Breunig        1.654681   
7963c2e6-8c7a-450d-a36f-12c9fa9a3354 2014   Montrezl Harrell      2.225606   
0af749be-c626-49f5-a558-aec6141a3421 2016   TJ Leaf               0.132213   
6a9269e3-b1b4-41b9-a8ab-df59886978be 2017   Lamar Stevens         2.796532   
346c6fc5-1a80-490b-a32b-d392d73de136 2017   Roland Griffin       -0.629022   
46688e90-022e-4190-b4c2-98b0af387562 2016   Terry Maston          0.512830   
2af2ce51-5235-4403-9841-4c9430cb8f1b 2015   Matt Costello         1.654681   
2fa0c978-1586-4201-a56f-093637ec3e10 2015   Marquese Chriss      -0.629022   
79ed938c-96e4-47e8-af2f-972d4009ffbc 2017   Averyl Ugba           1.083755   
0b3e3268-97a0-4815-9734-d31e06f1ad38 2016   Rashawn Thomas        3.253272   
b7e39c95-4504-426b-ba65-847d1ff61a27 2014   Scott Eatherton      -0.629022   
5b52856a-e7cc-43be-a039-73db14597d33 2016   David Collette        0.512830   
35156dfc-e3fa-4aa6-98ac-3d4744fe9cb5 2015   Brice Johnson         0.703138   
276050cc-348e-47df-b23b-574b5ddb4908 2013   T.J. Warren           1.083755   
30c49b9a-d7f7-4c35-8cd1-7f38babaa64a 2014   Twymond Howard        1.369218   
e8cb08f3-fc5c-4d71-83a1-31f2ddfa7d7a 2017   Kevin Knox            1.274064   
22106ca1-61b2-49b9-bf8d-00c52cc0cd13 2016   Jaron Blossomgame     0.512830   
7259bc8b-f7a0-49e4-b4a6-86f52d88e903 2015   Brandon Sherrod       0.512830   
12b20916-fa98-443e-9e9f-176d2f3d0359 2015   Jordan Washington    -0.629022   
1d98c7a7-1014-4d63-8061-4eb4920543da 2014   Eric Stutz            1.274064   
3d64c487-51d4-4d54-9d1a-e6e0a9e3d13b 2014   Josiah Heath          0.893447   
b50bf001-245d-4f61-b42e-020c05ec953a 2014   Zach Auguste         -0.058096   
5777f60a-7410-4650-aab4-ad4ed2719745 2017   Marvin Bagley III     0.798292   
ed2f8d67-3593-4b26-a242-a92bc114d740 2014   Frank Kaminsky        1.844989   
...                                                                    ...   
a5b2c9ba-1571-456c-9b33-b3f0bde96ef2 2013   Nick Banyard         -0.629022   
67e3101e-6e37-4fec-b663-e0dfac7c7a02 2013   Rodney Bullock       -0.629022   
905d462f-c797-4447-8f0e-78e1a2342c91 2017   Alou Dillon          -0.629022   
213b876c-0ad5-4788-9834-8a25fa6d0657 2016   Aaron Falzon         -0.629022   
c1079b6e-2408-4894-ac4e-380c010d432c 2016   Nate Grimes          -0.629022   
42661995-6077-406b-b4f5-befb0e55d726 2016   Dale Jones           -0.629022   
f47fbf0e-bdeb-4a7f-bff6-fcc3285d76ef 2013   Mogboluwaga Oginni   -0.629022   
9254f675-363d-418a-b087-6fc3a4127716 2016   Amadou Sidibe        -0.629022   
b832506a-4157-4e1b-9560-a5aa05bf4762 2016   John Carlson         -0.629022   
d4e48977-03a9-4536-b96f-73d3eecfc7f9 2016   AJ John              -0.629022   
12196f21-b5d7-4120-b05e-2418c6ad83c5 2017   Naji Ozeir           -0.629022   
7dbf2f42-b464-4eaa-b93c-1206e358c311 2016   Kyle Davis           -0.629022   
a97202bf-827e-411f-acc1-bce63f4d8fd0 2016   Landon Fuller        -0.629022   
669aa5c6-6da7-4c2e-ba9c-eef18bd20475 2015   Branden Sheppard     -0.629022   
a97202bf-827e-411f-acc1-bce63f4d8fd0 2017   Landon Fuller        -0.629022   
38d4396e-09c0-4e92-9fb4-f7e575882462 2013   Zac Grossenbacher    -0.629022   
072539cf-ede4-4100-9f7b-b90600f17435 2016   Alex Holcombe        -0.629022   
e8609814-fd06-4497-a28b-eec498e93722 2014   Davonte Fitzgerald   -0.629022   
8c4a956f-3976-4942-8baf-5071c807106e 2014   Leo Edwards          -0.629022   
6fe062f1-40

## 八村塁（Rui Hachimura）の NBA初年度成績を予測する

### 基本的なアイディア

大学時代に八村選手と同じようなスタッツを残した選手を見つけ出し、その選手が NBA 初年度に残した成績を参照することで、八村選手の活躍もある程度推測できるのではないか？

### データ

Google BigQuery に `bigquery-public-data/ncaa_basketball` という公開データがあるのでこれを使用する。
もともと kaggle で使用された課題用のデータだが、実際の NCAA のスタッツが実名で得られる。

```SQL
SELECT
    *
FROM
    `bigquery-public-data.ncaa_basketball.mbb_teams`
WHERE
    market="Gonzaga"
LIMIT 1;
```

として、チームID を取得したあとで、

```SQL
SELECT
    *
FROM
    `bigquery-public-data.ncaa_basketball.mbb_players_games_sr`
WHERE
    team_id="2f4d21f8-6d5f-48a5-abca-52a30583871a" AND
    first_name="Rui" AND
    tournament = "NCAA"
    LIMIT 1000;
```

とすると八村選手の登録されている試合の結果が取得できる。
BigQuery上で集計まで行っても良いが、できる限り python側に raw に近い状態で持ってきた方が扱いが楽なため、試合ごとのデータをダウンロードする。
なお、クエリ結果は 1GB までであればそのまま直接 json で Google Drive上に保存できる。

そして、まずは八村選手と身長、体重、ポジションの近いプレイヤーを抜き出す。

```SQL
SELECT
    *
FROM
     `bigquery-public-data.ncaa_basketball.mbb_players_games_sr`
 WHERE
     height BETWEEN 75 AND 85 AND
     weight BETWEEN 205 AND 245 AND
     position="F" AND
     season BETWEEN 2011 AND 2015 AND
     tournament = "NCAA"
LIMIT 100000;
```

BigQuery上では 2003年のデータから取得できるが、あまり昔のデータを引っ張ってきてもルールや文化が変わってしまっているので 2011年から 2015年に大学でプレーしていた選手に限定する。
また、身長は ±5-inch、体重は ±20-pounds とし、ポジションはフォワード（F）のプレイヤーだけを取得する。

ただ、BigQuery上にあるデータが 2017年までのものなので、八村選手がメインでプレーした 2018年のデータはない。
そこで、 ESPN から 2018年のデータを取得する。
https://www.espn.com/mens-college-basketball/team/stats/_/id/2250

実はこの ESPN から取得したデータは kaggle で用いられたものよりも粒度が荒く（とはいえバスケットボールで通常用いられる指標である）、このまま結合しようとすると色々と不都合があるのであとで再整形する。

### データ読み込み

BigQuery から取得したデータは jsonl形式（一行ずつが独立した jsonフォーマットになっているテキスト形式）なので、これを python で読み込んで pandas.DataFrame の形に変換する。

```python
import pandas as pd

DATA_PATH_RUI = './data/bq-results-20190710-040500-jn11labq5atn.json'
DATA_PATH_SIM = './data/bq-results-20190711-072404-txneu5c4yv68.json'


# jsonl を読み込んで pandas.DataFrame に変換する関数
def read_jsonl(path):
    with open(path, 'r') as f:
        data = f.readlines()
        
    data = map(lambda x: x.rstrip(), data)
    data_json_str = "[" + ','.join(data) + "]"
    return pd.read_json(data_json_str)


# 八村選手のデータを読み込む
df_rui = read_jsonl(DATA_PATH_RUI)
# 他の選手のデータを読み込む
df_sim = read_jsonl(DATA_PATH_SIM)
```

ちなみにこの時点で何人くらいの候補プレイヤーがいるかというと、

```python
names = df_sim.abbr_name.unique()
print(f'The number of similar players to Rui Hachimura = {len(names)}')
# => The number of similar players to Rui Hachimura = 1300
```

ということで 1,300人が該当した。
八村選手の身長は 2m を超え、体重は 100kg 近いが、5年間で同じような体格の選手がこれだけいるということである。
アメリカの大学バスケ市場の凄まじさがこれだけで伝わってくる。


### 前処理（preprocess）

さて、BigQuery から取得したデータは各試合ごとのスコア（ボックススコアと呼ばれる）なので、これを各年で集計する必要がある。
基本的にバスケットボールの世界では統計値には平均値のみを使用し、分散その他統計値はあまり用いられない。
ESPN のデータも例外ではなく平均値のみなので、平均を計算する関数で簡単に済ませられる。

さて、`DataFrame.mean()`メソッドを使用したいのだが、ここで注意しなければならない点がいくつかある。

1. 文字列は `mean()` によって無視される
2. 割合の平均は全体の割合に等しくない
3. 欠損値の扱い
4. （カラム毎のスケール）

今回使用するデータは非常に綺麗なフォーマットである。
綺麗とはどういうことかというと、文字列はちゃんと文字列として格納してあるし、数値は数値として格納してある、...ということである。
「何を当たり前のことを...？」と思う人もいるかもしれない。
それが、こと日本国内においてはこの当たり前でないデータが数多くある。

原理的に仕方ないものとしては、元号が一つの例として挙げられる。
元号は知っての通り、`令和元年` などと表記される。
この `元年` とは `1st year`　のことであり、当然 `2nd year` と連続する値である。
しかし、多くのソフトウェア、ミドルウェアでは `令和元年` と `令和2年` は連続する整数ではなく、ただの文字列として扱われる。
そこで、年次毎の売上推移を見たければ元号から西暦に変換する必要がある。

原理的に仕方無くないものに関しては、ここでは言及しない。
"ネ申エクセル" と揶揄されるものは大方これを含む。

話はそれたが、綺麗なフォーマットにおいては、`mean()` は正しく機能する。
何も考えずに `DataFrame.groupby(...).mean()` とすれば、`groupby()` した中での平均値を求めてくれる。
今回の例で言えば、`player_id`、`season`、`full_name` が固有なデータ行について集計してくれ、というものである。
多くの場合、`player_id` と `full_name` は一致するので、どちらか一方を指定すればよいのだが、あとでデータを見るときにどちらもあった方が便利なため残しておく。

ここまでで、綺麗なフォーマットであれば `mean()` で機械的に平均が計算できるとのべたが、**割合の平均は全体の割合に等しくない** という点に注意する必要がある。

例えば、一試合目で 2本打って 1本入り 50%、二試合目で 3本打って 2本入り 66.6..% とする。
このとき、割合の平均をとると、 `(50 + 66.6...) / 2 = 116.6... / 2= 58.3...` となる。
一方、全体では 5本打って 3本入ったのだから、`3 / 5 = 0.6 = 60%` である。
このように、割合の平均を取ると、全体の割合から歪んでしまうことに注意しなければならない。

最後に、欠損値について考えなければならない。
欠損値とはその名の通り、記録できていないデータである。
例えば、出場機会がなく得点が 0点のとき、そのまま 0 としてしまうとあとで困る（特に平均や除算を行う場合）ので、
NA や NaN といった疑似的な数値を入れることがある。
この欠損値をそのままにしても上手くいくアルゴリズムはいくつかある（[How to deal with Missing Value - xgboost](https://xgboost.readthedocs.io/en/latest/faq.html#how-to-deal-with-missing-value）)のだが、大抵はそうでない。
かといって、前述のようにただ 0 としてしまうのはあまり良くない。

そこで、「何もしない」、「平均値または中央値で埋める」、「最頻値で埋める」などといったヒューリスティックが使われている（[6 Different Ways to Compensate for Missing Values In a Dataset (Data Imputation with examples)](https://towardsdatascience.com/6-different-ways-to-compensate-for-missing-values-data-imputation-with-examples-6022d9ca0779)）。

しかし、これをすれば良いというものはなく、データセットや問題に応じてどの方法で欠損値を扱うかを毎回検討せねばならず、このことがデータサイエンティストが当たる最初の壁となる。

さらにもう一つ、処理するアルゴリズムによっては、カラムごとの値のスケールを調整しなければならない場合もある。
例えば、平均得点が 5点違うのと、平均ブロック数が 5つ違うのとでは、意味合いが全く異なる。
選手の類似性という点に着目すると決めている以上、この差異が十分よく反映されるように変換を施すべきである。

さて、前置きが長くなったが、今回の場合に照らすと

1. パーセンテージの平均は使わない（ちゃんと計算する）
2. 欠損値を埋める必要がある
3. 値を正規化、標準化する必要がある

となる。以下に具体的な処理を示す。

#### 使用するカラムを指定する

DataFrame を扱う際、カラムを指定して処理を行いたい場合が多いので、あらかじめ処理対象のカラムを記憶しておくと良い。
ここでは、`num_labels`という変数に数値型の値を格納している。

```python
str_labels = [
    # プレイヤーID
    'player_id',
    
    # プレイヤー名
    'full_name',

    # シーズン
    'season',    
]

num_labels = [
    # アシスト数
    'assists',

    # ブロック数
    'blocks',
    
    # ディフェンスリバウンド数
    'defensive_rebounds',

    # フィールドゴール試投数
    'field_goals_att',
    
    # フィールドゴール成功数
    'field_goals_made',
    
    # フリースロー試投数
    'free_throws_att',
    
    # フリースロー成功数
    'free_throws_made',
    
    # 身長
    'height',
    
    # 出場時間
    'minutes',
    
     # オフェンスリバウンド数
    'offensive_rebounds',
    
    # 得点数
    'points',
    
    # リバウンド数
    'rebounds',
    
    # スティール数
    'steals',
    
    # 3P試投数
    'three_points_att',
    
    # 3P成功数
    'three_points_made',
    
    # ターンオーバー数
    'turnovers',
    
    # 2P試投数
    'two_points_att',
    
    # 2P成功数
    'two_points_made',
    
    # 体重
    'weight',
]
```

#### 欠損値処理

今回は、欠損値はそのカラムの平均値で埋めるという処理を施す。

```python
# 値が数値か否かを判定するメソッド
def is_number(x):
    return (type(x) is int or type(x) is float) and not np.isnan(x)
```

```python
# 「値が数値ならそのままの値を、そうでなければ平均値を返すメソッド」を返すメソッド
def value_or_mean(df, label):
    mean = df[label].dropna().mean()

    def f(row):        
        v = None
        if is_number(row[label]):
             v = row[label]
        else:     
             v = mean
        return v
    return f


# 欠損値を埋める処理
def fillna(df):
    _df = df.copy()
    
    neighbors = []

    for label in num_labels:
        _df[label] = df.apply(value_or_mean(df, label), axis=1)
    return _df


df_sim = fillna(df_sim)
```

#### 統計処理

ここまで来て、やっと統計処理を行うことができる。

```python
# 八村選手のスタッツをシーズンごとに計算
df_rui_stats = df_rui.groupby(['player_id', 'season', 'full_name']).mean()

# 他のプレイヤーのスタッツをシーズンごとに計算
df_sim_stats = df_sim.groupby(['player_id', 'season', 'full_name']).mean()
```

#### 標準化

今回のケースでは、平均を引いて標準偏差で割る標準化を行う。
なぜ閉区間`[0, 1]` へのフラットな射影ではなく標準化をするかというと、今回重要なのは平均に比べてどのくらい異質なスコアを各プレイヤーが示しているかという点だからである。
例えば、平均得点が 30点のプレイヤーがいたとしよう。このとき、他のプレイヤーも同様に 30点近くを記録しているのであれば、そのプレイヤーは平均的な得点力、と見れるかも知れない。一方、他のプレイヤーの平均得点が高々10点近くなのであれば、そのプレイヤーの得点力は並外れていると評価できる。
このように、平均からどのくらい離れているか、異常であるかを示す変換として標準化を使用できる。

```python
def standardize(df, labels):
    _df = df.copy()
    for label in labels:
        # 平均と標準偏差を計算
        mean = df[label].mean()
        std = df[label].std()

        values = []
        for val in df[label].values:
            
            # 標準偏差が 0 でなければ標準化する
            if std > 0:
                values.append((val - mean) / std)
                
            # 標準偏差が 0 の場合は、どのプレイヤーも同じ値をとるということなので、
            # とりあえず 0 を入れておけば問題ない（厳密には平均を入れるが、類似度計算にはあまり関係ない）
            else:
                values.append(0.0)
                
        # ラベルに示されるカラムを更新
        _df[label] = values
    return _df

df = standardize(df, num_labels)
```

さて、ここまでが前処理である。
次節では、いよいよ類似度計算に移る。


### 類似度計算（Cosine類似度）

ある 2つのベクトルが類似しているとはどういうことだろうか？
例えば各ベクトルの大きさをとって、それぞれが近い値なら類似しているといえるかもしれない。
2つのベクトルを超平面上にプロットして、その距離が近くても類似しているといえるかもしれない。
さて、ここでの類似とは、「2人のプレイヤーが似たタイプのプレイを得意とする選手である」と定義する。
すなわち、似たプレイスタイルであれば、その強弱は考えない、ということである。
なぜこのように定義するかというと、NBA にドラフトされるような選手はその時点である程度プレイスタイルが完成されており、リーグに入ってもそれほど大きくプレイスタイルを変えないだろう、という知見による。
プレイスタイルが変わらないのであれば、異なるのはその洗練度であり、その選手がどう成長していくかも含めて参考になるだろう。

これらを踏まえ、類似度には Cosine類似度を採用する。
Cosine類似度は、ベクトルの内積をそれぞれの大きさの積で割ったものとして定義される。
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html

幾何学的には、2つのベクトルの角度が小さいほど、値が大きくなる指標である。
Cosine類似度は自分で実装しても良いのだが、python では `scikit-learn` に実装されているため、それを使用する。
このような指標（metric）は意外とハマるので極力ありものを使ったほうが良い。
ハマりどころは境界値（0 や inf）が多い。

```python
from sklearn.metrics.pairwise import cosine_similarity

def calc_cos(row, df):
    cosines = []
    for _, x in df.iterrows():
        X = row.values.reshape(1, -1)
        Y = x.values.reshape(1, -1)
        cosines.append(cosine_similarity(X, Y)[0][0])
    return cosines

# iloc[0] は八村選手のデータ
cos = calc_cos(df.iloc[0], df)

df['cosine'] = cos
```

Cosine類似度が計算できたら、後はソートするだけである。

```python
df_similar = df.sort_values('cosine', ascending=False)
```

この結果は、公開した jupyter notebook に載せているので興味があればご覧いただきたい。


## 結果

さて、以上より、Cosine類似度において類似していると思しき選手のリストを手に入れた。
そこで、類似度の高い選手について、NBA でどんな成績を残しているか見ていく。

ここで注意すべきなのは、Cosine類似度を用いているため、類似性はあくまでプレイスタイルに関してであり、実際にはそれほど活躍していない選手も載っているということだ。
そのような選手と、八村選手のようにリーグから注目されている選手とを区別するために、ドラフトされているか否かを指標に入れる。
ただし、このデータは配布されていないため、名前を Google検索して目でドラフト順位を記載したものである。
また、ドラフト外の選手は上述の理由で取り上げないこととする。

### 1st pick of Rui-like draft, Montrezl Harrell（Houston Rockets, 32nd）

まずは最もプレイスタイルが近いと判定された選手、Montrezl Harrell。
https://www.basketball-reference.com/players/h/harremo01.html

現在は LA Clippers に所属しており、カンファレンス・ファーストラウンドで昨季の王者である Golden State Warriors を苦しめた。

初年度成績はまずまずといったところ。
全体32位での指名なので、八村選手はこれよりも良いスタッツが期待できる。

エースとまではいかないが、若いチームで安定した数字を残しており、4年目でこのくらいの数字であれば NBA に残り続けることができるだろう、という成績だ。
ちなみに契約金は 600万ドル（およそ7億円）と報じられている。
https://www.espn.com/nba/team/roster/_/name/lac/la-clippers

ただし今シーズンは大型選手補強が行われており、プレイタイムは減るかも知れない。


#### 初年度成績（Houston Rockets）

* 39ゲーム出場（1ゲーム先発）
* 平均9.7分プレイ
* 平均3.6得点
* 平均1.7リバウンド
* 平均0.3スティール
* 平均0.3ブロック
* 平均0.4ターンオーバー
* 平均1.2パーソナルファウル


#### 最新シーズン成績（4年目：LA Clippers）

* 82ゲーム出場（5ゲーム先発）
* 平均26.3分プレイ
* 平均16.6得点
* 平均6.5リバウンド
* 平均0.9スティール
* 平均1.3ブロック
* 平均1.6ターンオーバー
* 平均3.1パーソナルファウル

### 2nd pick of Rui-like draft, T.J. Leaf（Indiana Pacers, 18th）

イスラエル出身のプレイヤー T.J. Leaf。
https://www.basketball-reference.com/players/l/leaftj01.html

八村選手と同じく 1ラウンドでの指名（18位）であるが、正直なところ、それほど目立つ活躍はできていない。
趣旨が変わってしまうのであまり詳しくは調査していないが、もしかするとコミュニケーションの問題があったのかもしれない。

#### 初年度成績（Indiana Pacers）

* 53ゲーム出場（0ゲーム先発）
* 平均8.7分プレイ
* 平均2.9得点
* 平均1.5リバウンド
* 平均0.1スティール
* 平均0.1ブロック
* 平均0.2ターンオーバー
* 平均0.8パーソナルファウル


#### 最新シーズン成績（2年目：Indiana Pacers）

* 58ゲーム出場（1ゲーム先発）
* 平均9.0分プレイ
* 平均3.9得点
* 平均2.2リバウンド
* 平均0.2スティール
* 平均0.3ブロック
* 平均0.2ターンオーバー
* 平均0.6パーソナルファウル

### 3rd pick of Rui-like draft, Marquese Chriss（Sacramento Kings, 8th）

全体8位指名で、今回最も八村選手と指名順が近い Marquese Chriss選手。
https://www.basketball-reference.com/players/c/chrisma01.html

初年度はなんと全ゲームに出場、内75ゲームで先発するという期待のされ方。
また、スタッツも初年度にしては非常に良く、上々のスタートだったが、現在はトレードされて出場機会もぐっと減ってしまっている。


#### 初年度成績（Sacramento Kings）

* 82ゲーム出場（75ゲーム先発）
* 平均21.3分プレイ
* 平均9.2得点
* 平均4.2リバウンド
* 平均0.8スティール
* 平均0.9ブロック
* 平均1.3ターンオーバー
* 平均3.2パーソナルファウル

#### 最新シーズン成績（5年目：Cleveland Cavaliers）

* 27ゲーム出場（2ゲーム先発）
* 平均14.6分プレイ
* 平均5.7得点
* 平均4.2リバウンド
* 平均0.6スティール
* 平均0.3ブロック
* 平均0.9ターンオーバー
* 平均2.4パーソナルファウル


### 考察

チームの育成、起用方針に大きく左右されるので一概には言えないが、実力や期待度的には Marquese Chriss と同じくらいのプレイタイムが与えられ、成績を残してもおかしくない。
一方で、やはりそれほど甘くない世界、成績が伸び悩めばトレード、最悪解雇されることもありうる。
八村選手は渡邊選手とともに日本の期待を大きく背負っているが、直近の成績を重視せず、 Montrezl Harrell のように成長を続けて欲しい。

### 結論（予想）

Washinton Wizards の事情も考えて、だいたい Marquese Chriss の上位互換となるような成績を残すのではないだろうか。

* 70ゲーム出場（56ゲーム先発）
* 平均25.0分プレイ
* 平均14.0得点
* 平均6.0リバウンド
* 平均0.8スティール
* 平均2.0ブロック
* 平均1.3ターンオーバー
* 平均3.2パーソナルファウル